lib/src/json/builder/builder.dart

```diff
@@ -0,0 +1,184 @@
+/// ```dart
+/// var optionalJsonInt = jsonOptional(jsonInt);
+/// var optionalInt = optionalJsonInt(reader);  // int or null
+/// ```
+JsonBuilder<T /*?*/ > jsonOptional<T>(JsonBuilder<T> builder) =>
+    (JsonReader reader) => reader.tryNull() ? null : builder(reader);
+
+/// Reads an array of values from a [JsonReader].
+///
+/// Reads an array from the provided `reader`, where each
+/// element of the array is built by [elementBuilder].
+JsonBuilder<List<T>> jsonArray<T>(JsonBuilder<T> elementBuilder) =>
+    (JsonReader reader) {
+      reader.expectArray();
+      var result = <T>[];
+      while (reader.hasNext()) {
+        result.add(elementBuilder(reader));
+      }
+      return result;
+    };
+
+/// Reads an array of values from [JsonReader].
+///
+/// The builder used for each element is provided
+/// by the [elementBuilder] function based on the
+/// index of the element. This allows, for example,
+/// building an array that alternates between two
+/// types.
+JsonBuilder<List<T>> jsonIndexedArray<T>(
+        JsonBuilder<T> elementBuilder(int index)) =>
+    (JsonReader reader) {
+      reader.expectArray();
+      var result = <T>[];
+      var index = 0;
+      while (reader.hasNext()) {
+        result.add(elementBuilder(index)(reader));
+        index++;
+      }
+      return result;
+    };
+
+/// Builds a value from a JSON array.
+///
+/// Builds a value for each element using [elementBuilder],
+/// then folds those into a single value using [initialValue]
+/// and [combine], just like [Iterable.fold].
+JsonBuilder<T> jsonFoldArray<T, E>(JsonBuilder<E> elementBuilder,
+        T initialValue(), T combine(T previus, E elementValue)) =>
+    (JsonReader reader) {
+      reader.expectArray();
+      var result = initialValue();
+      while (reader.hasNext()) {
+        var element = elementBuilder(reader);
+        result = combine(result, element);
+      }
+      return result;
+    };
+
+/// Builds a map from a JSON object.
+///
+/// Reads a JSON object and builds a value for each object value
+/// using [valueBuilder]. Then creates a `Map<String, T>` of
+/// the keys and built values.
+JsonBuilder<Map<String, T>> jsonObject<T>(JsonBuilder<T> valueBuilder) =>
+    (JsonReader reader) {
+      reader.expectObject();
+      var result = <String, T>{};
+      String key;
+      while ((key = reader.nextKey()) != null) {
+        result[key] = valueBuilder(reader);
+      }
+      return result;
+    };
+
+/// Builds a map from a JSON object
+///
+/// Reads a JSON object and builds a value for each object value
+/// using the builder returned by `valueBuilders[key]` for the corresponding
+/// object key.
+/// Then creates a `Map<String, T>` of the keys and built values.
+///
+/// The [defaultBuilder] is used if [valueBuilders] has no entry for
+/// a give key. If there is no [defaultBuilder] and not entry in
+/// [valueBuilders] for a key, then the entry is ignored.
+JsonBuilder<Map<String, T>> jsonStruct<T>(
```

Comment:
It is likely common to not care about the keys you haven't at all. Should we have an additional boolean option here `skipUnspecifiedKeys` or some such?

Reply:
That's what a missing `defaultBuilder` will do.